lib/runtime-core/src/parse.rs

```diff
@@ -110,11 +110,46 @@ pub fn read_module<
     let mut namespace_builder = Some(StringTableBuilder::new());
     let mut name_builder = Some(StringTableBuilder::new());
     let mut func_count: usize = 0;
+    let mut mcg_signatures_fed = false;
     let mut mcg_info_fed = false;
 
     loop {
         use wasmparser::ParserState;
         let state = parser.read();
+
+        // Feed signature and namespace information as early as possible.
```

Comment:
Maybe it makes sense to use the Wasmparser types that let us be more specific about what exactly we want: https://docs.rs/wasmparser/0.51.4/wasmparser/struct.CodeSectionReader.html .  They have a bunch of section-specific parsers, there aren't validating versions of these parsers though, so we'd have to do a validation pass first

Reply:
Since this block borrows `mcg`, there would be lifetime issues if this is made into a closure. And yeah this is about code reuse. The same code was already duplicated twice for `BeginFunctionBody` and `EndWasm`, and this change will add one more duplication if done "in place".

Reply:
`mcg` could be passed as an argument here to make closures work